server/http_oauth2_test.go

```diff
@@ -197,3 +203,176 @@ func TestEndpoints_OAuth2_Options(t *testing.T) {
+		{"code, wrong azp claim", "07_couper.hcl", "/cb?code=qeuboub-wazp-id", http.Header{"Cookie": []string{"nnc=" + st}}, http.StatusForbidden, "", "", "access control error: ac: azp claim / client ID mismatch, azp = 'bar', client ID = 'foo'"},
+		{"code, missing iss claim", "07_couper.hcl", "/cb?code=qeuboub-miss-id", http.Header{"Cookie": []string{"nnc=" + st}}, http.StatusForbidden, "", "", "access control error: ac: token issuer is invalid: 'iss' value doesn't match expectation"},
+		{"code, wrong iss claim", "07_couper.hcl", "/cb?code=qeuboub-wiss-id", http.Header{"Cookie": []string{"nnc=" + st}}, http.StatusForbidden, "", "", "access control error: ac: token issuer is invalid: 'iss' value doesn't match expectation"},
+		{"code, wrong aud claim", "07_couper.hcl", "/cb?code=qeuboub-waud-id", http.Header{"Cookie": []string{"nnc=" + st}}, http.StatusForbidden, "", "", "access control error: ac: token audience is invalid: 'foo' wasn't found in aud claim"},
+		{"code; client_secret_basic; PKCE", "04_couper.hcl", "/cb?code=qeuboub", http.Header{"Cookie": []string{"pkcecv=qerbnr"}}, http.StatusOK, "code=qeuboub&code_verifier=qerbnr&grant_type=authorization_code&redirect_uri=http%3A%2F%2Flocalhost%3A8080%2Fcb", "Basic Zm9vOmV0YmluYnA0aW4=", ""},
+		{"code; client_secret_post", "05_couper.hcl", "/cb?code=qeuboub", http.Header{"Cookie": []string{"pkcecv=qerbnr"}}, http.StatusOK, "client_id=foo&client_secret=etbinbp4in&code=qeuboub&code_verifier=qerbnr&grant_type=authorization_code&redirect_uri=http%3A%2F%2Flocalhost%3A8080%2Fcb", "", ""},
+		{"code, state param", "06_couper.hcl", "/cb?code=qeuboub&state=" + state, http.Header{"Cookie": []string{"st=" + st}}, http.StatusOK, "code=qeuboub&grant_type=authorization_code&redirect_uri=http%3A%2F%2Flocalhost%3A8080%2Fcb", "Basic Zm9vOmV0YmluYnA0aW4=", ""},
+		{"code, nonce param", "07_couper.hcl", "/cb?code=qeuboub-id", http.Header{"Cookie": []string{"nnc=" + st}}, http.StatusOK, "code=qeuboub-id&grant_type=authorization_code&redirect_uri=http%3A%2F%2Flocalhost%3A8080%2Fcb", "Basic Zm9vOmV0YmluYnA0aW4=", ""},
+	} {
+		t.Run(tc.path[1:], func(subT *testing.T) {
+			shutdown, hook := newCouper("testdata/oauth2/"+tc.filename, test.New(t))
+			defer shutdown()
+
+			req, err := http.NewRequest(http.MethodGet, "http://back.end:8080"+tc.path, nil)
+			helper.Must(err)
+
+			for k, v := range tc.header {
+				req.Header.Set(k, v[0])
+			}
+			req.Header.Set("X-AS-Origin", oauthOrigin.URL)
+
+			res, err := client.Do(req)
+			helper.Must(err)
+
+			if res.StatusCode != tc.status {
+				t.Errorf("%q: expected Status %d, got: %d", tc.name, tc.status, res.StatusCode)
+			}
+
+			tokenResBytes, err := ioutil.ReadAll(res.Body)
+			var jData map[string]interface{}
+			json.Unmarshal(tokenResBytes, &jData)
+			if params, ok := jData["form_params"]; ok {
+				if params != tc.params {
+					t.Errorf("%q: expected params %s, got: %s", tc.name, tc.params, params)
```

Comment:
```suggestion
					subT.Errorf("%q: expected params %s, got: %s", tc.name, tc.params, params)
```

applies to all `t`'s .